### ISSUE TITLE
Retrieving the last inserted ObjectId or Filename

### ISSUE THREAD

**Issue by Andargor** [devconcept/multer-gridfs-storage#3]
I would like to retrieve the ObjectId (or even filename) from the callback, and this even if there was an error. At least, I have not figured out how to do so. To illustrate, this is what I am trying to achieve.

I would like to have efficient, implicit deduplication when a document is inserted in GridFS. My root name is "attachments", for this example. To do so, I have created a unique index on the "md5" field in the attachment.files collection. This means that there should be an error if a filename with the same md5 hash is inserted. And I can then handle it, although I don't know how to retrieve the ObjectID or md5 hash from this transaction.

There is a problem with this: when the document is inserted, It is first created in attachments.chunks, and then hashed, and only then the attachment.files entry is attempted to be created, and raises the error if there's a duplicate. Which means there are dangling chunks with no corresponding files entry. Again, if I could know what the ObjectId was at the moment the attempt was made, I could handle it and simply remove the duplicate chunks. How do I do this?

My workaround is a naive approach where I manually md5 hash the file before and checking if it exists before attempting to insert, but that seems to me inefficient, since it will be hashed a second time.

**Comment by devconcept** [comment]
You are right. Your current design is flawed and this is because, not only you have duplicated work, but also because the index is now sitting there doing nothing.  Instead of trying to figure out which was the last filename or ObjectId you should let the index do its work and handle the error on the way back.

For this I added the [`error`](https://github.com/devconcept/multer-gridfs-storage#event-error) event in version 1.2.1. This events only triggers when a streaming error is found, in your case the index preventing the insertion of the document. This event has two arguments; the first is the error itself that will help you to determine if it was an insertion error or something else and the second is the file configuration. You can use the second argument to figure out the filename of the failed upload. 

If you still need an ObjectId you can generate one yourself using the identifier option. Something like this will do
```javascript
var ObjectID = require(‘mogodb’). ObjectID;
var storage = new GridFsStorage({
   …
   identifier: function(req, file, cb) {
      cb(null, new ObjectID());
   }
});
```
Basically, the ObjectId is generated by the underlying stream if none is provided so the only way to know which was generated is after a successful upload. Using your own identifier will allow you to catch the error knowing the generated id beforehand.

You should note that if a file fails to upload for any reason the whole request is considered invalid and any other file that was uploaded in the same request will be deleted as well. That part is handled by the multer module and is by design. You should only worry about this if you are implementing multiple uploads in a single request.

Let me know if it worked for you so I can close this issue.

**Comment by Andargor** [comment]
I disagree with your assessment on the approach, but I had not considered you could recreate the storage on every insert to have a predictable id that you could then handle on an error from a duplicate file and erase the dangling chunks.

In any case, your example answers the original request and in the meantime I've implemented the "naive" approach of checking the md5 before an insert, so I have no objection for you to close.

**Comment by devconcept** [comment]
You disagree but I'm still not clear why. If you don't want to use the index why add it in the first place. Remember that multer calls this plugin and its configuration functions **per file** and you **can** have more than one per upload (eg. [array](https://github.com/expressjs/multer#arrayfieldname-maxcount), [fields](https://github.com/expressjs/multer#fieldsfields) or [any](https://github.com/expressjs/multer#any)). This makes "remembering" an id somehow problematic and error prone. Can you show me a fragment of code to ilustrate your case.

**Comment by devconcept** [comment]
I wrote a small test and this code should work for you

```javascript
const storage = new GridFsStorage({
    // your config here
    identifier: function(req, file, cb) {
        cb(null, mongo.ObjectID());
    }
});

storage.on('error', (err, config) => {
    // Make sure the error matches 
    if (err.name === 'MongoError' && err.code === 11000) {
        db.collection('attachments.chunks')
            .deleteOne({files_id: config._id});
    }
});
```

This will take care of the dangling chunks. Honestly, I think this can't get any easier. 

I'm closing this issue for now.
